File: src/main/java/com/refinedmods/refinedstorage/block/FluidStorageBlock.java
package com.refinedmods.refinedstorage.block;

import com.refinedmods.refinedstorage.apiimpl.network.node.storage.FluidStorageNetworkNode;
import com.refinedmods.refinedstorage.apiimpl.storage.FluidStorageType;
import com.refinedmods.refinedstorage.container.FluidStorageContainerMenu;
import com.refinedmods.refinedstorage.container.factory.BlockEntityMenuProvider;
import com.refinedmods.refinedstorage.blockentity.FluidStorageBlockEntity;
import com.refinedmods.refinedstorage.util.BlockUtils;
import com.refinedmods.refinedstorage.util.NetworkUtils;
import net.minecraft.core.BlockPos;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.phys.BlockHitResult;
import net.minecraftforge.network.NetworkHooks;

public class FluidStorageBlock extends NetworkNodeBlock {
    private final FluidStorageType type;

    public FluidStorageBlock(FluidStorageType type) {
        super(BlockUtils.DEFAULT_ROCK_PROPERTIES);

        this.type = type;
    }

    public FluidStorageType getType() {
        return type;
    }

    @Override
    public void setPlacedBy(Level level, BlockPos pos, BlockState state, LivingEntity player, ItemStack stack) {
        if (!level.isClientSide) {
            FluidStorageNetworkNode storage = ((FluidStorageBlockEntity) level.getBlockEntity(pos)).getNode();

            if (stack.hasTag() && stack.getTag().hasUUID(FluidStorageNetworkNode.NBT_ID)) {
                storage.setStorageId(stack.getTag().getUUID(FluidStorageNetworkNode.NBT_ID));
            }

            storage.loadStorage(player instanceof Player ? (Player) player : null);
        }

        // Call this after loading the storage, so the network discovery can use the loaded storage.
        super.setPlacedBy(level, pos, state, player, stack);
    }

    @Override
    public BlockEntity newBlockEntity(BlockPos pos, BlockState state) {
        return new FluidStorageBlockEntity(type, pos, state);
    }

    @Override
    @SuppressWarnings("deprecation")
    public InteractionResult use(BlockState state, Level level, BlockPos pos, Player player, InteractionHand handIn, BlockHitResult hit) {
        if (!level.isClientSide) {
            return NetworkUtils.attemptModify(level, pos, player, () -> NetworkHooks.openGui((ServerPlayer) player, new BlockEntityMenuProvider<FluidStorageBlockEntity>(
                ((FluidStorageBlockEntity) level.getBlockEntity(pos)).getNode().getTitle(),
                (blockEntity, windowId, inventory, p) -> new FluidStorageContainerMenu(blockEntity, player, windowId),
                pos
            ), pos));
        }

        return InteractionResult.SUCCESS;
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/block/FluidStorageBlock.kt
package com.refinedmods.refinedstorage.block

import com.refinedmods.refinedstorage.RS
//import com.refinedmods.refinedstorage.apiimpl.network.node.storage.FluidStorageNetworkNode
import com.refinedmods.refinedstorage.apiimpl.storage.FluidStorageType
//import com.refinedmods.refinedstorage.tile.FluidStorageTile
import com.refinedmods.refinedstorage.tile.NoOpBlockEntity
import com.refinedmods.refinedstorage.util.BlockUtils
import com.thinkslynk.fabric.annotations.registry.RegisterBlock
import net.minecraft.block.BlockEntityProvider
import net.minecraft.block.BlockState
import net.minecraft.block.entity.BlockEntity
import net.minecraft.entity.LivingEntity
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.item.ItemStack
import net.minecraft.util.ActionResult
import net.minecraft.util.Hand
import net.minecraft.util.hit.BlockHitResult
import net.minecraft.util.math.BlockPos
import net.minecraft.world.BlockView
import net.minecraft.world.World

open class FluidStorageBlock(val type: FluidStorageType):
        NetworkNodeBlock(BlockUtils.DEFAULT_ROCK_PROPERTIES, false) // TODO Double check connected value
//        BlockEntityProvider
{
    companion object {
        const val SIXTY_FOUR_K_FLUID_STORAGE_BLOCK_ID = "64k_fluid_storage_block"
        const val TWO_HUNDRED_FIFTY_SIX_K_FLUID_STORAGE_BLOCK_ID = "256k_fluid_storage_block"
        const val THOUSAND_TWENTY_FOUR_K_FLUID_STORAGE_BLOCK_ID = "1024k_fluid_storage_block"
        const val FOUR_THOUSAND_NINETY_SIX_K_FLUID_STORAGE_BLOCK_ID = "4096k_fluid_storage_block"
        const val CREATIVE_FLUID_STORAGE_BLOCK_ID = "creative_fluid_storage_block"
    }

//    override fun createBlockEntity(world: BlockView?): BlockEntity?
//            = NoOpBlockEntity()
    // TODO BlockEntities
//            = FluidStorageTile(type)

    // TODO Network
//    override fun onPlaced(world: World, pos: BlockPos, state: BlockState, placer: LivingEntity?, itemStack: ItemStack) {
//        if (!world.isClient) {
//            (world.getBlockEntity(pos) as FluidStorageTile?)?.node?.let {
//                itemStack.tag?.let { tag ->
//                    if (tag.containsUuid(FluidStorageNetworkNode.NBT_ID)) {
//                        it.storageId = tag.getUuid(FluidStorageNetworkNode.NBT_ID)
//                    }
//                    it.loadStorage()
//                }
//            }
//        }
//
//        // Call this after loading the storage, so the network discovery can use the loaded storage.
//        super.onPlaced(world, pos, state, placer, itemStack)
//    }

    override fun onUse(state: BlockState, world: World, pos: BlockPos, player: PlayerEntity, hand: Hand, hit: BlockHitResult): ActionResult {
        // TODO Port Gui
//        return if (!world.isClient) {
//            NetworkUtils.attemptModify(world, pos, hit.getFace(), player) {
//                NetworkHooks.openGui(player as ServerPlayerEntity?, PositionalTileContainerProvider<FluidStorageTile>(
//                        (world.getBlockEntity(pos) as FluidStorageTile?)!!.node.title,
//                        { tile: FluidStorageTile?, windowId: Int, inventory: PlayerInventory?, p: PlayerEntity? -> FluidStorageContainer(tile, player, windowId) },
//                        pos
//                ), pos)
//            }
//        } else ActionResult.SUCCESS
        return ActionResult.SUCCESS
    }
}

@RegisterBlock(RS.ID, FluidStorageBlock.SIXTY_FOUR_K_FLUID_STORAGE_BLOCK_ID)
class SixtyFourKFluidStorageBlock: FluidStorageBlock(FluidStorageType.SIXTY_FOUR_K)
@RegisterBlock(RS.ID, FluidStorageBlock.TWO_HUNDRED_FIFTY_SIX_K_FLUID_STORAGE_BLOCK_ID)
class TwoHundredFiftySixKFluidStorageBlock: FluidStorageBlock(FluidStorageType.TWO_HUNDRED_FIFTY_SIX_K)
@RegisterBlock(RS.ID, FluidStorageBlock.THOUSAND_TWENTY_FOUR_K_FLUID_STORAGE_BLOCK_ID)
class ThousandFourKFluidStorageBlock: FluidStorageBlock(FluidStorageType.THOUSAND_TWENTY_FOUR_K)
@RegisterBlock(RS.ID, FluidStorageBlock.FOUR_THOUSAND_NINETY_SIX_K_FLUID_STORAGE_BLOCK_ID)
class FourThousandNinetySixKFluidStorageBlock: FluidStorageBlock(FluidStorageType.FOUR_THOUSAND_NINETY_SIX_K)
@RegisterBlock(RS.ID, FluidStorageBlock.CREATIVE_FLUID_STORAGE_BLOCK_ID)
class CreativeFluidStorageBlock: FluidStorageBlock(FluidStorageType.CREATIVE)